### PR TITLE
Make TableCache fully thread-safe

### PR DESCRIPTION
**Pull request by gervandiepen**
Issue #217 Make TableCache fully thread-safe

### CODE REVIEW

File: tables/Tables/PlainTable.cc [hunk 1]
```diff
@@ -285,12 +285,19 @@ void PlainTable::closeObject()
...
+        //# seems to be asynchronous to some degree, so try a few times.
+        int nTrys = 5;
+        while (isMultiUsed(False)) {
+            if (nTrys == 0) {
+                unmarkForDelete (False, "");
+                throw (TableError ("Table " + name_p + " cannot be deleted;"
+                                   " the table or a subtable is still used"
+                                   " in another process"));
+            }
+            sleep(1);
```
Comment (juliantaylor) [inline tables/Tables/PlainTable.cc]: sleep has poor interaction with signals, please use nanosleep

Comment (gervandiepen) [inline tables/Tables/PlainTable.cc]: Thanks. Good point.
I got the fix from somebody else.

File: tables/Tables/TableCache.cc [hunk 2]
```diff
@@ -47,19 +57,6 @@ PlainTable* TableCache::operator() (const String& tableName) const
...
-    ScopedMutexLock sc(itsMutex);
-    return (PlainTable*) (tableMap_p.getVal (index));
-}
-
-uInt TableCache::ntable() const
-{
-    ScopedMutexLock sc(itsMutex);
-    return tableMap_p.ndefined();
-}
-
```
Comment (juliantaylor) [inline tables/Tables/TableCache.cc]: removal of these two functions is not backward compatible and broke CASA compilation, I'll check if it is easy to fix within casa.

Comment (juliantaylor) [inline tables/Tables/TableCache.cc]: was easy, casa build should be fixed.

Comment (gervandiepen) [inline tables/Tables/TableCache.cc]: Great, thanks.

File: tables/Tables/TableCache.cc [hunk 3]
```diff
@@ -95,5 +92,113 @@ void TableCache::rename (const String& newName, const String& oldName)
...
+}
+
+Vector<String> TableCache::getTableNames() const
+{
+    ScopedMutexLock sc(itsMutex);
+    uInt ntab = tableMap_p.ndefined();
+    Vector<String> names;
+    for (uInt i=0; i<ntab; i++) {
+	PlainTable& table = *static_cast<PlainTable*>(tableMap_p.getVal(i));
+        names[i] = table.tableName();
```
Comment (juliantaylor) [inline tables/Tables/TableCache.cc]: does this work for casa::Vector? shouldn'T this be push_back?

**Pull request merged**